src/core/thread/topology.cpp

```diff
@@ -111,6 +111,84 @@ void Neighbor::GenerateChallenge(void)
         Random::Crypto::FillBuffer(mValidPending.mPending.mChallenge, sizeof(mValidPending.mPending.mChallenge)));
 }
 
+Child::AddressIterator::AddressIterator(const Child &aChild, otChildIp6AddressIterator aIterIndex)
+    : mChild(aChild)
+    , mFilter(Ip6::Address::kTypeAny)
+{
+    // `aIterIndex` value of zero indicates start (or mesh-local IPv6
+    // address). Non-zero value specifies the index into address array
+    // starting from one for first element (i.e, `aIterIndex - 1` gives
+    // the array index).
+
+    Action action = kInit;
+
+    if (aIterIndex != OT_CHILD_IP6_ADDRESS_ITERATOR_INIT)
+    {
+        mAddress = &mChild.mIp6Address[aIterIndex - 1];
+        action   = kCheck;
+    }
+
+    Update(action);
+}
+
+otChildIp6AddressIterator Child::AddressIterator::ConvertToChildIp6AddressIterator(void) const
+{
+    otChildIp6AddressIterator iterIndex;
+
+    if (mAddress == &mMeshLocalAddress)
+    {
+        iterIndex = 0;
```

Comment:
What about save the iter index as `mIterIndex` and then just return it. The code here would be much simpler. 
**The index is actually useful for MLR features because it's used to index in a bit mask.**

Reply:
Good suggestion.  
It makes the `GetAddress()` a bit more complex, but can simplify the rest of the methods. Overall I think would be good. Will change it later.

Reply:
Changed in new pushed commit.